### PR TITLE
EMA Decay Rate Sweep: 0.999→{0.9995, 0.9998} — More Averaging for OOD Generalization

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current EMA model uses `--ema_decay 0.999`, which gives an effective memory of ~1/(1-0.999) = 1000 gradient steps. With ~77 steps/epoch × 160 epochs = ~12,320 total steps, the EMA "sees" about the last 8% of training.

The EMA model is the final evaluation model — it IS the prediction we report. A higher decay value averages over more training history, which has two effects:
1. **Smoother parameters** — less sensitivity to random gradient noise near the end of training
2. **More flat-minima sampling** — higher decay effectively averages across a longer trajectory segment, which biases toward flat loss basin regions (better OOD generalization)

This is directly motivated by the SWA (Stochastic Weight Averaging) literature. Our prior `--swad` attempt was catastrophic because it averaged throughout all of training; this approach uses the existing EMA mechanism but with slower decay (more averaging of the final phase of training).

**Expected impact:** p_tan -0.5% to -1.5%, especially p_oodc and p_re (OOD metrics that benefit most from smoother predictions). No parameter count change, no architecture change.

## Instructions

Run 4 experiments: 2 decay values × 2 seeds. Use `--wandb_group phase6/ema-decay-sweep`.

```bash
# decay=0.9995, seed 42
cd cfd_tandemfoil && python train.py --agent nezuko \
  --wandb_name "nezuko/ema-decay9995-s42" --wandb_group phase6/ema-decay-sweep \
  --ema_decay 0.9995 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15

# decay=0.9995, seed 73
cd cfd_tandemfoil && python train.py --agent nezuko \
  --wandb_name "nezuko/ema-decay9995-s73" --wandb_group phase6/ema-decay-sweep \
  --ema_decay 0.9995 --seed 73 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15

# decay=0.9998, seed 42
cd cfd_tandemfoil && python train.py --agent nezuko \
  --wandb_name "nezuko/ema-decay9998-s42" --wandb_group phase6/ema-decay-sweep \
  --ema_decay 0.9998 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15

# decay=0.9998, seed 73
cd cfd_tandemfoil && python train.py --agent nezuko \
  --wandb_name "nezuko/ema-decay9998-s73" --wandb_group phase6/ema-decay-sweep \
  --ema_decay 0.9998 --seed 73 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15
```

**Note:** No `--aft_foil_srf_context` (known guard bug, pending frieren #2134 fix). The flag `--ema_decay` already exists in Config and argparse.

**Watch for:** EMA warmup behavior — with decay=0.9998, the EMA converges more slowly at startup (ema_start_epoch=140 still applies). This should be fine since the EMA averaging epoch range is the same (epochs 140-160).

Report 2-seed avg for each decay value and W&B run IDs for all 4 runs.

## Baseline

Current single-model baseline (PR #2119, PCGrad 2-way):

| Metric | 2-seed target |
|--------|--------------|
| p_in   | < 13.20      |
| p_oodc | < 7.91       |
| p_tan  | **< 29.48**  |
| p_re   | < 6.50       |

Baseline W&B: jpe1t13t (s42, ema_decay=0.999), cdccuyl7 (s73, ema_decay=0.999)

```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results — EMA Decay Rate Sweep

### Setup
- 4 runs: 2 decay values (0.9995, 0.9998) × 2 seeds (42, 73)
- W&B group: `phase6/ema-decay-sweep`
- All runs completed ~150-151 epochs (180-min timeout)
- No code changes — pure hyperparameter sweep

### Individual Run Results (Surface MAE — Pressure)

| Run | W&B ID | Decay | Seed | p_in | p_oodc | p_tan | p_re |
|-----|--------|-------|------|------|--------|-------|------|
| ema-decay9995-s42 | `gadyfr02` | 0.9995 | 42 | 13.016 | 7.596 | 28.786 | 6.547 |
| ema-decay9995-s73 | `cts1sttc` | 0.9995 | 73 | 13.246 | 7.800 | 29.931 | 6.637 |
| ema-decay9998-s42 | `cwi47jtn` | 0.9998 | 42 | 13.423 | 7.878 | 29.176 | 6.795 |
| ema-decay9998-s73 | `b3t8ob3d` | 0.9998 | 73 | 13.219 | 7.947 | 30.044 | 6.725 |

### 2-Seed Means vs Baseline (p_in < 13.20, p_oodc < 7.91, p_tan < 29.48, p_re < 6.50)

| Variant | p_in | p_oodc | p_tan | p_re |
|---------|------|--------|-------|------|
| **decay=0.9995 mean** | **13.131 (-0.5%)** ✅ | **7.698 (-2.7%)** ✅ | **29.359 (-0.4%)** ✅ | 6.592 (+1.4%) ❌ |
| decay=0.9998 mean | 13.321 (+0.9%) ❌ | 7.913 (~0%) | 29.610 (+0.4%) ❌ | 6.760 (+4.0%) ❌ |
| Baseline (0.999) | 13.20 | 7.91 | 29.48 | 6.50 |

### Analysis

**decay=0.9995 is a clear winner — beats baseline on 3 of 4 metrics.**

1. **decay=0.9995** improves p_in (-0.5%), p_oodc (-2.7%), and p_tan (-0.4%) while only slightly regressing p_re (+1.4%). The p_oodc improvement is particularly strong and consistent across both seeds.

2. **decay=0.9998 is too slow** — it averages too aggressively and loses model sharpness. All metrics regress or stay flat. The effective memory of 5000 steps (vs 2000 for 0.9995 and 1000 for 0.999) is too long for our ~12K total step budget.

3. **Seed 42 with decay=0.9995** (`gadyfr02`) is the standout run: p_in=13.016, p_oodc=7.596, p_tan=28.786 — all well below baseline targets.

4. **The SWA connection holds**: moderate additional averaging (0.9995 vs 0.999) smooths the EMA trajectory enough to improve OOD generalization without over-smoothing.

5. **Val loss**: decay=0.9995 has the best val losses (0.383-0.388), while 0.9998 is worse (0.390-0.391), confirming 0.9995 is the sweet spot.

### Suggested Follow-ups
1. **Recommend merge with decay=0.9995** — the improvement is clean and free (no parameter count change, no architecture change, no compute cost increase).
2. An 8-seed validation would confirm the -2.7% p_oodc improvement is statistically significant.
3. The p_re regression is mild (+1.4%) and within seed variance — may not be systematic.

**Comment by tcapelle** [comment]
ADVISOR: Good result — decay=0.9995 beats the OLD baseline on 3/4 metrics. The p_oodc -2.7% improvement is particularly strong.

However, **the baseline just moved.** PR #2130 (GSB + PCGrad) merged with p_tan=28.60, adding `--gap_stagger_spatial_bias` to the baseline command. Your runs don't include this flag, so we need a rebased validation.

## New Baseline (PR #2130)

| Metric | New target |
|--------|-----------|
| p_in   | < 13.05   |
| p_oodc | < 7.70    |
| p_tan  | **< 28.60** |
| p_re   | < 6.55    |

## Instructions

Rebase onto current noam:
```bash
git fetch origin noam
git rebase origin/noam
```

Run **2 seeds with decay=0.9995** (the winner) on the new baseline:
```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent nezuko \
  --wandb_name "nezuko/ema9995-gsb-s42" --wandb_group phase6/ema-decay-gsb \
  --ema_decay 0.9995 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Seed 73 (same, --seed 73, wandb_name "nezuko/ema9995-gsb-s73")
```

If 2-seed mean p_tan < 28.60, this merges immediately.

**Comment by tcapelle** [comment]
## Revision Results — decay=0.9995 + GSB Baseline

### Changes
- Rebased onto current noam (includes PR #2130 `--gap_stagger_spatial_bias`)
- Running decay=0.9995 (the round 1 winner) with full new baseline flags

### Results (Surface MAE — Pressure)

| Run | W&B ID | Seed | p_in | p_oodc | p_tan | p_re |
|-----|--------|------|------|--------|-------|------|
| ema9995-gsb-s42 | `mnbrh0r2` | 42 | 13.333 | 7.874 | 29.789 | 6.544 |
| ema9995-gsb-s73 | `kptwsdvt` | 73 | 13.706 | 7.669 | 29.704 | 6.692 |

### vs New Baseline (PR #2130: p_in < 13.05, p_oodc < 7.70, p_tan < 28.60, p_re < 6.55)

| Metric | New Baseline | decay=0.9995+GSB mean | Delta |
|--------|-------------|----------------------|-------|
| p_in | 13.05 | 13.519 | +3.6% ❌ |
| p_oodc | 7.70 | 7.772 | +0.9% ❌ |
| p_tan | 28.60 | 29.746 | +4.0% ❌ |
| p_re | 6.55 | 6.618 | +1.0% ❌ |

### Analysis

**Does not beat the new baseline on any metric.**

The decay=0.9995 advantage observed in round 1 (vs the old baseline without GSB) does not carry over when combined with `--gap_stagger_spatial_bias`. The GSB feature likely changes the loss landscape in a way that makes slower EMA averaging less beneficial — possibly because GSB already provides the spatial smoothing that higher EMA decay was approximating.

The p_tan regression (+4.0%) is particularly notable — round 1 showed p_tan improvement with 0.9995, but the GSB baseline is already so strong on p_tan (28.60) that the EMA change can't compete.

**Recommendation:** Close this PR. The decay=0.9995 idea was sound and worked against the pre-GSB baseline, but the GSB merge obsoleted the advantage.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2141 because ADVISOR: Closing — decay=0.9995 doesn't compound with GSB. All metrics regress vs new baseline (p_tan +4.0%, p_in +3.6%).

Your analysis is correct: GSB already provides the spatial smoothing that higher EMA decay was approximating. The GSB baseline (p_tan=28.60) is a much tougher target than the pre-GSB baseline where 0.9995 showed promise.

**Confirmed:** ema_decay=0.999 remains optimal for the current architecture.

New assignment coming shortly..
